internal/providers/qemu/qemu_fwcfg.go

```diff
@@ -34,22 +38,74 @@ import (
 )
 
 func FetchConfig(f *resource.Fetcher) (types.Config, report.Report, error) {
+	// load qemu_fw_cfg module
 	_, err := f.Logger.LogCmd(exec.Command("modprobe", "qemu_fw_cfg"), "loading QEMU firmware config module")
 	if err != nil {
 		return types.Config{}, report.Report{}, err
 	}
 
-	data, err := ioutil.ReadFile(firmwareConfigPath)
+	// get size of firmware blob, if it exists
+	sizeBytes, err := ioutil.ReadFile(firmwareConfigSizePath)
 	if os.IsNotExist(err) {
 		f.Logger.Info("QEMU firmware config was not found. Ignoring...")
+		return util.ParseConfig(f.Logger, []byte{})
 	} else if err != nil {
-		f.Logger.Err("couldn't read QEMU firmware config: %v", err)
+		f.Logger.Err("couldn't read QEMU firmware config size: %v", err)
+		return types.Config{}, report.Report{}, err
+	}
+	size, err := strconv.Atoi(strings.TrimSpace(string(sizeBytes)))
+	if err != nil {
+		f.Logger.Err("couldn't parse QEMU firmware config size: %v", err)
 		return types.Config{}, report.Report{}, err
 	}
 
+	// Read firmware blob.  We need to make as few, large read() calls as
+	// possible, since the qemu_fw_cfg kernel module takes O(offset)
+	// time for each read syscall.  ioutil.ReadFile() would eventually
+	// converge on the correct read size (one page) but we can do
+	// better, and without reallocating.
+	// Leave an extra guard byte to check for EOF
+	data := make([]byte, 0, size+1)
+	fh, err := os.Open(firmwareConfigPath)
+	if err != nil {
+		f.Logger.Err("couldn't open QEMU firmware config: %v", err)
+		return types.Config{}, report.Report{}, err
+	}
+	defer fh.Close()
+	lastReport := time.Now()
+	reporting := false
+	for len(data) < size {
+		// if size is correct, we will never call this at an offset
+		// where it would return io.EOF
+		n, err := fh.Read(data[len(data):cap(data)])
+		if err != nil {
+			f.Logger.Err("couldn't read QEMU firmware config: %v", err)
+			return types.Config{}, report.Report{}, err
+		}
+		data = data[:len(data)+n]
+		if !reporting && time.Since(lastReport).Seconds() >= 10 {
+			f.Logger.Warning("Reading QEMU fw_cfg takes quadratic time. Consider moving large files or config fragments to a remote URL.")
+			reporting = true
+		}
+		if reporting && (time.Since(lastReport).Seconds() >= 5 || len(data) >= size) {
+			f.Logger.Info("Reading config from QEMU fw_cfg: %d/%d KB", len(data)/1024, size/1024)
+			lastReport = time.Now()
+		}
+	}
+	if len(data) > size {
+		// overflowed into guard byte
+		f.Logger.Err("missing EOF reading QEMU firmware config")
+		return types.Config{}, report.Report{}, errors.New("missing EOF")
+	}
+	// If size is not at a page boundary, we know we're at EOF because
+	// the guard byte was not filled.  If size is at a page boundary,
+	// trust that firmwareConfigSizePath was telling the truth to avoid
+	// incurring an extra read call to check for EOF.  We're at the end
+	// of the file so the extra read would be maximally expensive.
```

Comment:
If we're not at a page boundary, we're not trusting `firmwareConfigSizePath` because an overrun will write into the guard byte and be caught by the `len(data) > size` check.  And underruns will be caught in any case by `fh.Read` returning `io.EOF`.  It's only at a page boundary that we're trusting `firmwareConfigSizePath` not to be smaller than reality.

Reply:
Gotcha, I see what you mean. The confusing thing is that `firmwareConfigSizePath` informs the buffer size in the first place. So I think it's more like we're trusting `firmwareConfigSizePath` and *not* trusting `firmwareConfigPath`? But I'm just being pedantic at this point. :)